In [1]:
import numpy as np
import pandas as pd
import os

# 1.1 Call sklearn libraries

# 1.1.1 Split data into train and test data
from sklearn.model_selection import train_test_split

# 1.1.2 PReprocessong:
from sklearn.preprocessing import StandardScaler

# 1.1.3 Import class DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier as dtree

# 1.1.4 To draw decision tree
from sklearn.tree import export_graphviz

# 1.1.5
from sklearn.tree import plot_tree
import matplotlib.pyplot as plt

In [2]:
import seaborn as sns

In [4]:
df = pd.read_csv("C:\\Users\\Dell\\Downloads\\archive\\Real_fake_news.csv")

In [5]:
df.head()

Unnamed: 0  Unnamed: 0.1               author  \
0           0             0         Jill Terreri   
1           1             1  Eleanor Hildebrandt   
2           2             2     Emily Zirkelbach   
3           3             3        Madeline Heim   
4           4             4        Sabine Martin   

                                           statement  \
0  New York is "leading the nation with the lowes...   
1  “Iowa ranks 45th in the U.S. for internet conn...   
2  “FUN FACT: West Virginia is home to zero billi...   
3  Says gubernatorial candidate Rebecca Kleefisch...   
4  The Biden administration "published a study co...   

                                              source  target BinaryTarget  \
0  NYS Department of Corrections and Community Su...    True         REAL   
1                                         Cindy Axne    True         REAL   
2                                         David Gura    True         REAL   
3                                   Kelda Helen Roys    True         REAL   
4                                     Randy Feenstra    True         REAL   

   BinaryNumTarget  
0                1  
1                1  
2                1  
3                1  
4                1

In [6]:
df.columns = list(map(lambda a: a.lstrip(), df.columns))

In [7]:
df.isnull().sum()

Unnamed: 0         0
Unnamed: 0.1       0
author             0
statement          0
source             0
target             0
BinaryTarget       0
BinaryNumTarget    0
dtype: int64

In [8]:
df.shape

(4870, 8)

In [9]:
df['author'].value_counts()

Ciara O'Rourke        669
Tom Kertscher         311
Louis Jacobson        250
Samantha Putterman    242
Jon Greenberg         153
                     ... 
Magdaline Duncan        1
Ryan Hanrahan           1
Asjylyn Loder           1
Allison Garfield        1
Brian Grace             1
Name: author, Length: 318, dtype: int64

In [10]:
df = df.drop(columns=['Unnamed: 0','Unnamed: 0.1','target','BinaryTarget'])

In [11]:
df_final = pd.get_dummies(df)
df_final.head()

BinaryNumTarget  author_Aaron Marshall  author_Aaron Sharockman  \
0                1                      0                        0   
1                1                      0                        0   
2                1                      0                        0   
3                1                      0                        0   
4                1                      0                        0   

   author_Adam Thorp  author_Adrianne Uphold  author_Adriel Bettelheim  \
0                  0                       0                         0   
1                  0                       0                         0   
2                  0                       0                         0   
3                  0                       0                         0   
4                  0                       0                         0   

   author_Al Bernardi  author_Alan Gathright  author_Alan Hovorka  \
0                   0                      0                    0   
1                   0                      0                    0   
2                   0                      0                    0   
3                   0                      0                    0   
4                   0                      0                    0   

   author_Alayna Degenhardt  ...  source_WrongWayRubio.com  \
0                         0  ...                         0   
1                         0  ...                         0   
2                         0  ...                         0   
3                         0  ...                         0   
4                         0  ...                         0   

   source_Xavier Becerra  source_Yard sign  \
0                      0                 0   
1                      0                 0   
2                      0                 0   
3                      0                 0   
4                      0                 0   

   source_Yes On 62 campaign to abolish California's death penalty  \
0                                                  0                 
1                                                  0                 
2                                                  0                 
3                                                  0                 
4                                                  0                 

   source_YouTube videos  source_YourContent  source_Yvette McGee Brown  \
0                      0                   0                          0   
1                      0                   0                          0   
2                      0                   0                          0   
3                      0                   0                          0   
4                      0                   0                          0   

   source_Zach Wahls  source_Zerlina Maxwell  source_teaparty.org  
0                  0                       0                    0  
1                  0                       0                    0  
2                  0                       0                    0  
3                  0                       0                    0  
4                  0                       0                    0  

[5 rows x 6678 columns]

In [13]:
X = df_final.drop(columns=['BinaryNumTarget'])
y = df_final['BinaryNumTarget']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.30, random_state=42)

In [14]:
ct = dtree(
           criterion="entropy",    # Alternative 'entropy'
           max_depth=None       # Alternative, specify an integer
                              # 'None' means full tree till single leaf
           )
# 5.1
_=ct.fit(X_train,y_train)

In [15]:
y_te = ct.predict(X_test)
np.sum((y_test == y_te))/y_test.size

0.8870636550308009

In [16]:
fi = ct.feature_importances_
fi

array([0., 0., 0., ..., 0., 0., 0.])

In [17]:
from sklearn.ensemble import RandomForestClassifier # using random forest 

In [18]:
#Create a Gaussian Classifier
clf=RandomForestClassifier(n_estimators=100)

#Train the model using the training sets y_pred=clf.predict(X_test)
clf.fit(X_train,y_train)

y_pred=clf.predict(X_test)

In [19]:
#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics
# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.8836413415468857


When using decision tree , the accuracy was 88% , while using the Random Forest Classifier accuracy has increased to 90%